File: jbake-core/src/main/java/org/jbake/util/PagingHelper.java
package org.jbake.util;

import java.net.URI;
import java.net.URISyntaxException;

public class PagingHelper {
    private static final String URI_SEPARATOR = "/";
    private long totalDocuments;
    private int postsPerPage;

    public PagingHelper(long totalDocuments, int postsPerPage) {
        this.totalDocuments = totalDocuments;
        this.postsPerPage = postsPerPage;
    }

    public int getNumberOfPages() {
        return (int) Math.ceil((totalDocuments * 1.0) / (postsPerPage * 1.0));
    }

    public String getNextFileName(int currentPageNumber) throws URISyntaxException {
        if (currentPageNumber < getNumberOfPages()) {
            return new URI((currentPageNumber + 1) + URI_SEPARATOR).toString();
        } else {
            return null;
        }
    }

    public String getPreviousFileName(int currentPageNumber) throws URISyntaxException {

        if (isFirstPage(currentPageNumber)) {
            return null;
        } else {
            if ( currentPageNumber == 2 ) {
                // Returning to first page, return empty string which when prefixed with content.rootpath should get to root of the site.
                return "";
            }
            else {
                return new URI((currentPageNumber - 1) + URI_SEPARATOR).toString();
            }
        }
    }

    private boolean isFirstPage(int page) {
        return page == 1;
    }

    public String getCurrentFileName(int page, String fileName) throws URISyntaxException {
        if ( isFirstPage(page) ) {
            return fileName;
        }
        else {
            return new URI(page + URI_SEPARATOR + fileName).toString();
        }
    }

}

File: jbake-core/src/main/kotlin/org/jbake/util/PagingHelper.kt
package org.jbake.util

import java.net.URI
import java.net.URISyntaxException
import kotlin.math.ceil

class PagingHelper(totalDocuments: Long, postsPerPage: Int) {

    val numberOfPages: Int = ceil((totalDocuments * 1.0) / (postsPerPage * 1.0)).toInt()

    /** Returns the next page file name or null if there is no next page.  */
    @Throws(URISyntaxException::class)
    fun getNextFileName(currentPage: Int): String? =
        if (currentPage >= this.numberOfPages) null
        else validated("${currentPage + 1}/")


    @Throws(URISyntaxException::class)
    fun getPreviousFileName(currentPage: Int): String?
        = when (currentPage) {
            1 -> null
            // Returning to first page -> return "" -> prefixed with `content.rootpath` should get to site root.
            2 -> ""
            else -> validated("${currentPage - 1}/")
        }

    @Throws(URISyntaxException::class)
    fun getCurrentFileName(page: Int, fileName: String): String =
        if (page == 1) fileName
        else validated("$page/$fileName")

    private fun validated(path: String) = URI(path).toString()
}
